dask/array/linalg.py

```diff
@@ -675,9 +675,13 @@ def compression_matrix(data, q, n_power_iter=0, seed=None, compute=False):
         state = seed
     else:
         state = RandomState(seed)
+    if (data.dtype).itemsize <= 4:
```

Comment:
I think it would probably make more sense if this maintained consistency with numpy linalg dtype preservation (at least for `svd` and `qr`), which would mean anything other than float32, complex64 or complex128 would become float64.

Perhaps:

```suggestion
    datatype = data.type
    if datatype not in {np.float32, np.complex64, np.complex128}:
        datatype = np.float64
```

I don't know why numpy does it that way, but I can see value in dask svd/svd_compressed and numpy svd having the same semantics.

Reply:
@eric-czech, if I am right LAPACK does not provide routines for 16 bit or 128 bit calculations (not sure either whether numpy/scipy provides support for those type towards LAPACK). So numpy/scipy automatically promotes it to float32 when it is lower than 32 bits. It keeps complex datatypes, but everything else gets translate into float I believe. The same will be true for the QR decomposition I believe. For complex input data types it is comes naturally to sample its columns following the same precision, but with non-complex weights. Don't see why this doesn't happen for np.complex32. 

It is a good suggestion to maintain consistency with numpy! I will review this on Wednesday (11/11).

Reply:
@eric-czech: so reviewing your suggestion. First of all, I couldn't find the suggestion you made in numpy linalg, but in sklearn they have something similar to your suggestion so I presume you meant sklearn. We should keep in mind that we are sampling a matrix columns of the "data" variable, with a certain dtype. We want to make sure that that dtype doesn't get promoted to a higher one. Therefore, I have made a small experiment to see how numpy reacts on different datatypes:

```python
import numpy as np

types = {'int', 'uint', 'float', 'complex'}
bits = {'int': {'8', '16', '32', '64'},
        'uint': {'8', '16', '32', '64'},
        'float': {'32', '64'},
        'complex': {'64', '128'}}

x = np.random.rand(100,10)
for dt in types:
    for bit in bits[dt]:
        u, s, vt = np.linalg.svd(x.astype(dt+bit, copy=False), full_matrices=False)
        print(dt+bit+':', u.dtype, s.dtype, vt.dtype)
        del(u, s, vt)
```

to my surprise this is the outcome (same behaviour for single chunk and multiple chunks dask svd and qr):

> complex64: complex64 float32 complex64
> complex128: complex128 float64 complex128
> int8: float64 float64 float64
> int16: float64 float64 float64
> int32: float64 float64 float64
> int64: float64 float64 float64
> uint8: float64 float64 float64
> uint16: float64 float64 float64
> uint32: float64 float64 float64
> uint64: float64 float64 float64
> float32: float32 float32 float32
> float64: float64 float64 float64

Complex input is what to be expected, float input is what is to be expected, only thing that is remarkable is that all integers types are all converted to float64. I am not sure whether this is a LAPACK thingy or rather a numpy decision. And although the impact might be little for in-memory calculations, blowing up your int8 matrix to float64 might be dramatic if you are not aware of it, already float32 is times 4, if a reconstruction of the matrix is desired of course. 

Nevertheless, if we want to maintain the numpy standard, then everything gets promoted to float64, EXCEPT for complex64 (which is float32 under the hood) and float32 itself. **sklearn makes a mistake by also promoting complex64 via float64**, as the reconstructed matrix will become complex128 to compensate for the float64 singular values:

```python
import numpy as np

u = np.random.rand(100,10).astype('complex64', copy=False)
s_32 = np.diag(np.random.rand(10,10)).astype('float32', copy=False)
s_64 = np.diag(np.random.rand(10,10)).astype('float64', copy=False)
vt = np.random.rand(10,100).astype('complex64', copy=False)

a = (u*s_32)@vt
b = (u*s_64)@vt

print(a.dtype, b.dtype)
```

> complex64 complex128

Since we are randomly sampling, we only want non-complex weights, such that only float32 or float64 can be the dtype of our sampling matrix omega. Simply said we want our datatype to be

**float32** in case input matrix is: **float32 or complex64**
**float64** in case input matrix is: **everything else**.

I will to find out how to make a proposal for this :p:.

Reply:
> It keeps complex datatypes, but everything else gets translate into float I believe. The same will be true for the QR decomposition I believe

Nice, good to know.

> First of all, I couldn't find the suggestion you made in numpy linalg, but in sklearn they have something similar to your suggestion so I presume you meant sklearn

I meant numpy.linalg.svd though I wasn't referring to code, just my own loop I ran at some point in the past like you ran checking output dtypes based on input dtypes.  I didn't know that "sklearn makes a mistake by also promoting complex64 via float64" and honestly hadn't thought that much about how complex types should get handled.

> Simply said we want our datatype to be
> float32 in case input matrix is: float32 or complex64
> float64 in case input matrix is: everything else.

Sounds good to me, thanks for digging into it more!